### DISCUSSION TITLE
Tutorials?

### DISCUSSION THREAD

**Discussion by Sorry-Im-Stupid** [bnpr/Malt#109] [General]
Since I've managed to get Malt mostly working, thanks to @pragma37 , I'm now playing around with actual textures, and node trees.
And I now have the difficult task of figuring out what all these nodes _actually do._ I've been perusing youtube looking for any kind of explanation of the malt nodes and node systems, and finding almost nothing. I would also like to know if there's any way to build a texture tree in eevee and translate/link it to the Malt node tree.
I'd really appreciate if anyone could point me to anything like this, but if it doesn't exist, perhaps we can build an infographic here to explain these. If I figure it out, I'd be happy to make a series of tutorials on youtube over time, for future noobs. 
Thoughts?

**Comment by pragma37** [comment]
This is the only tutorial for Malt nodes, AFAIK:
https://youtu.be/tE99jgCCcNE

And I'll keep posting examples in #94.

Aside from that, you can always get my latest Malt scene on [Patreon](https://www.patreon.com/pragma37) and the previous ones on [Gumroad](https://app.gumroad.com/pragma37).
I want to update them to use nodes, but all of them are code-only at the moment. 

This is an aspect where I hope to get some help from the community.
I'm pretty bad at making tutorials and it takes me a long time to record them.

That said, if you have any specific question about how to do something, open a thread in the Q&A section and I'll do my best to solve it.

**Reply by Sorry-Im-Stupid** [reply]
Yeah that's the one! The one explanation I've found anywhere.
It helped a lot in getting a structure that puts pixels on the screen.

But as it is, I can't get it to do much else, and the control I have over the end result is limited. Rather than a series of specific questions, a general quick reference of the functions available, and how they link together is my goal. Otherwise answering my questions is all you'll be doing for the next few months. I don't like eating people's time like that if I can avoid it. Which is why I was hoping for more of a community participation towards the goal of making an infographic reference sheet.

If I had to ask one question, at the moment, it would be: Is there a way to link an Eevee shader tree to a Malt node tree? If I can do that, then making a noob-level tutorial is within my capacity.

**Reply by Sorry-Im-Stupid** [reply]
For example, I've made this in eevee:
![toon base eevee3](https://user-images.githubusercontent.com/84992987/128197233-9a728c82-6c6c-4f79-b56a-1571ae4a10c8.png)
And I like the shadow detail for the most part, but it's too _smooth_, for lack of a better word.
Meanwhile in Malt, I have managed to build a node tree that does this:
![Test mat Malt2](https://user-images.githubusercontent.com/84992987/128197541-f80a40c7-124d-4d31-9cb5-59af859a10e2.jpg)
Which only looks like this if I export in .jpg format, for some reason. 

But if I could combine these materials I would get the shadow detail I want, with the roughness and outline style I like.
That's the kind of tutorial I'd like to be making as I go.

**Reply by pragma37** [reply]
Long story short, there's no way to automatically use EEVEE nodes in Malt.

**Reply by Sorry-Im-Stupid** [reply]
I've tried making a matcap in eevee and using it in malt, and it kinda works, but it stretches any details inside the circle, making it look weird. 
I know I can't use an eevee structure inside malt, but surely there's a way to create a procedural equation in eevee and link the _output_ such that malt can use it? I really only need voronoi, noise, and some math.

**Reply by Sorry-Im-Stupid** [reply]
I've made some progress on the style I want, with a matcap made in eevee, and some random math in Malt to make the noise I've been looking for. It's actually a bit more noise than I want, so if anyone can tell me how to turn it down just a bit, and get those fireflies out of the dark areas, I'd appreciate it.
Here's the render, matcap, and the node tree:
![test eevee matcap malt shader](https://user-images.githubusercontent.com/84992987/128932039-1b60fd09-c902-4960-888c-05317e97ea39.png)
![eevee base test matcap3](https://user-images.githubusercontent.com/84992987/128932122-487cc979-8332-4aa8-8c25-41984316186e.png)
![malt matcap noise node tree](https://user-images.githubusercontent.com/84992987/128932181-556d027f-043b-43f1-ba25-2ed2beb44905.png)

**Reply by Sorry-Im-Stupid** [reply]
I figured out how to tame the noise, just put a clamp between the get_random, and the float_multiply.
I was reminded of another peculiarity, how the image doesn't keep the noise effect very well if I save it in .png mode. It works perfectly in .jpg.
![test eevee matcap malt shader3](https://user-images.githubusercontent.com/84992987/129072908-a1a123fd-f89f-4ef4-9053-203e90224627.png)
![test eevee matcap malt shader3](https://user-images.githubusercontent.com/84992987/129072931-b7208ccd-711a-430e-acb9-6295b07f07af.jpg)

**Reply by pragma37** [reply]
Oh my! There are many wrong things there. 😅

1. You are mixing the fractional part of a number and the tangent of another one, but mixing with a factor of 1 means you are just using the second input.
2. You are using it as the seed for get_random, but you don't need any special setup for that, just leave it at 0.
3. I have no idea what's your rationale for using the BRDF_specular_cook_torrance like that, but that's super wrong. Maybe you are looking for get_specular ?
**I would recommend you to stick to the Common, Node Utils and NPR MeshShader nodes. Everything else is only useful for advanced use cases.**
4. Shadow color is just for transparent shadows (this needs a better name).
5. I'm not sure you understood the vec4_from_vec3 function. "v" is the RGB part and "a" is the alpha.
6. Are you actually using the extra Outputs? Those are available from the RenderLayer input, for post-processing.

> I know I can't use an eevee structure inside malt, but surely there's a way to create a procedural equation in eevee and link the output such that malt can use it? I really only need voronoi, noise, and some math.

I would bake the textures instead of using matcaps. There's currently no voronoi or cell noise in Malt but I plan to add them soon.

Also, please, try to make one thread per question and keep them as specific as possible, so it's easier for other users to search for already answered questions.

**Reply by Sorry-Im-Stupid** [reply]
Well, interesting about the random seed. Is there a reason that even has an input, if that's how it works?
I use the cook_torrance node because it's more adjustable than the regular specular.
And you're right, I don't understand how any of these things work. That's why I want to make a proper informational.
But the PixelOutput thing only takes a vec4 into color, so...
I don't know what these are supposed to do, but even individually, none of them do anything but dark grey to black. The colors don't show at all.
![Specular what](https://user-images.githubusercontent.com/84992987/129221713-e6597c4a-8fdb-4991-a301-e1304c1bd1e0.png)

**Reply by pragma37** [reply]
> Well, interesting about the random seed. Is there a reason that even has an input, if that's how it works?

If you need more than one random number (per pixel and per sample) in your shader, you use different seeds for each get_random node, but you can just use 0, 1, 2... and so on. It also allows you tweak the seed until you get the random result you want.

> But the PixelOutput thing only takes a vec4 into color, so...

So you set the alpha either to 1 or 0 (or enable transparency if that's what you want).

> I don't know what these are supposed to do, but even individually, none of them do anything but dark grey to black. The colors don't show at all.

Again, the ShadingModels function are for advanced stuff.
http://graphicrants.blogspot.com/2013/08/specular-brdf-reference.html
http://www.codinglabs.net/article_physically_based_rendering_cook_torrance.aspx
